File: Lucy/2025/3-W03/1161_Maximum_Level_Sum_of_a_Binary_Tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

// 두 번째 풀이: BFS(모범답안) + level 한 번에 찾기

function maxLevelSum(root: TreeNode | null): number {
  let queue = [root];

  let maxSum = -Infinity;
  let minLevel = 0;
  let level = 0;

  while (queue.length) {
    const tempQueue: TreeNode[] = [];
    let sum = 0;

    for (let i = 0; i < queue.length; i++) {
      const node = queue[i];

      if (node.left) tempQueue.push(node.left);
      if (node.right) tempQueue.push(node.right);

      sum += node.val;
    }

    queue = tempQueue;
    level++;

    if (sum > maxSum) {
      maxSum = sum;
      minLevel = level;
    }
  }

  return minLevel;
}

// 첫 번째 풀이: BFS + 포인터 + level 분리해서 찾기
function maxLevelSum(root: TreeNode | null): number {
  const queue: [TreeNode | null, number][] = [];
  queue.push([root, 0]);

  let front = 0;

  const levelSums: number[] = [];
  while (front < queue.length) {
    const [node, level] = queue[front++];

    if (!node) continue;

    if (level <= levelSums.length && !levelSums[level]) levelSums.push(0);

    levelSums[level] += node.val;

    if (node.left) queue.push([node.left, level + 1]);
    if (node.right) queue.push([node.right, level + 1]);
  }

  const maxSum = Math.max(...levelSums);
  for (let i = 0; i < levelSums.length; i++) {
    const currentSums = levelSums[i];
    if (maxSum === currentSums) return i + 1;
  }

  return -1;
}

File: Seori/LeetCode 75/Binary Tree_BFS/1161. Maximum Level Sum of a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        answer = 1
        max_sum = root.val
        depth, now_sum = 1, root.val
        
        # BFS 탐색을 각 depth마다 할 수 있도록 하여 최대 합계를 구함
        queue = [root]
        while queue:
            len_depth = len(queue)
            for _ in range(len_depth):
                node = queue.pop(0)
                now_sum += node.val
                if node.left is not None:
                    queue.append(node.left)
                if node.right is not None:
                    queue.append(node.right)

            if max_sum < now_sum:
                max_sum = now_sum
                answer = depth
            depth += 1
            now_sum = 0
        
        return answer